credentials/ssocreds/provider.go

```diff
@@ -0,0 +1,176 @@
+package ssocreds
+
+import (
+	"context"
+	"crypto/sha1"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/internal/sdk"
+	"github.com/aws/aws-sdk-go-v2/service/sso"
+)
+
+const ProviderName = "SSOProvider"
+
+var defaultCacheLocation = filepath.Join(getHomeDirectory(), ".aws", "sso", "cache")
+
+// GetRoleCredentialsAPIClient is a API client that implements the GetRoleCredentials operation.
+type GetRoleCredentialsAPIClient interface {
+	GetRoleCredentials(ctx context.Context, params *sso.GetRoleCredentialsInput, optFns ...func(*sso.Options)) (*sso.GetRoleCredentialsOutput, error)
+}
+
+// Options is the Provider options structure.
+type Options struct {
+	Client GetRoleCredentialsAPIClient
+
+	// The AWS account that is assigned to the user.
+	AccountID string
+
+	// The region where the AWS Signle Sign-On (AWS SSO) user portal is hosted.
+	Region string
+
+	// The role name that is assigned to the user.
+	RoleName string
+
+	// The URL that points to the organization's AWS Single Sign-On (AWS SSO) user portal.
+	StartURL string
+}
+
+// Provider is an AWS credential provider that retrieves temporary AWS credentials by exchanging an SSO login token.
+type Provider struct {
+	options Options
+}
+
+// New returns a new AWS Signle Sign-On (AWS SSO) credential proivder.
+func New(client GetRoleCredentialsAPIClient, accountID, region, roleName, startURL string, optFns ...func(options *Options)) *Provider {
```

Comment:
Are accountID, region, ... all required, and thus explicit params? 

Can we diverge from the pattern of constructors not returning an error, and have validation here for missing required members instead of explicit params?

Wouldn't this constructor break if SSO spec adds a new Option in the future, and that is made a required param? (Though that may never happen)

Reply:
Yes these are all required parameters, and match the ones that must be provided in the `[profile]` section for the SSO configuration. Our pattern for the providers has been to have the required parameters declared on the type constructor function, and then optional values would be provided via the function options passed in.

I don't think new required parameters could be added to the profile section as to do so would be a backwards incompatible change. Likewise any new parameters would have to be optional and thus this constructor design would be fine.

Also you could make this statement about any service like STS where we have providers that have a series of required and optional parameters so not sure what additional thought paths you were going for.

Reply:
Agree using function parameters to describe required-ness is a more obvious means of communicating that concept. It is possible that future versions of SSO come out that don't need all of these parameters, but in that case an alternative function would probably be the best answer. I'd not expect that to happen often.

Reply:
Right, I didn't think about the usage with STS.  Optional params that are added later, can be used with functional options, or any WithHelpers we expose.